compiler/src/iree/compiler/Dialect/LinalgExt/Utils/MatchUtils.cpp

```diff
@@ -0,0 +1,316 @@
+  return res;
+}
+
+/// Infer the iterator types from the init affine map. This looks at which dims
+/// are present in the map results, and returns an iterator types array with
+/// parallel types for dims that are present, and reduction types for dims that
+/// are not present.
+static FailureOr<SmallVector<utils::IteratorType>>
+inferIteratorsFromOutMap(AffineMap map) {
+  if (!map.isProjectedPermutation()) {
+    return failure();
+  }
+  SmallVector<utils::IteratorType> iterators(map.getNumDims(), red);
+  for (auto expr : map.getResults()) {
+    if (auto dim = dyn_cast<AffineDimExpr>(expr)) {
+      iterators[dim.getPosition()] = par;
+    }
+  }
+  return iterators;
+}
+
+bool isScaledContractionBody(Block &block) {
+  if (block.empty() || !block.back().mightHaveTrait<OpTrait::IsTerminator>()) {
+    llvm::errs() << "no terminator in the block";
+    return false;
+  }
+  if (block.getNumArguments() != 5) {
+    llvm::errs() << "expected block with 3 arguments";
+    return false;
+  }
+
+  Operation *terminator = block.getTerminator();
+  if (terminator->getNumOperands() != 1) {
+    llvm::errs() << "expected terminator with 1 operand";
+    return false;
+  }
+
+  // Skip cast-like ops and ops with no memory effects
+  auto getSourceSkipIrrelevant = [](Value value) -> Value {
+    Operation *op = value.getDefiningOp();
+    while (op) {
+      if ((op->getNumOperands() == 1) ||
+          !isa<arith::ExtFOp, arith::TruncFOp, arith::ScalingExtFOp,
+               arith::ScalingTruncFOp>(op)) {
+        break;
+      }
+      auto iface = dyn_cast<MemoryEffectOpInterface>(op);
+      if (!iface || !iface.hasNoEffect()) {
+        break;
+      }
+      value = op->getOperand(0);
+      op = value.getDefiningOp();
+    }
+    return value;
+  };
+  Value yielded = getSourceSkipUnary(terminator->getOperand(0));
+  Operation *reductionOp = yielded.getDefiningOp();
+  if (reductionOp->getNumResults() != 1 || reductionOp->getNumOperands() != 2) {
+    llvm::errs() << "expected reduction op to be binary";
+    return false;
+  }
+
+  Value reductionLHS = getSourceSkipUnary(reductionOp->getOperand(0));
+  Value reductionRHS = getSourceSkipUnary(reductionOp->getOperand(1));
+
+  if (reductionLHS != block.getArgument(4) &&
+      reductionRHS != block.getArgument(4)) {
+    llvm::errs()
+        << "expected reduction to take block argument #4 as one of the "
+           "operands (modulo unary casts)";
+    return false;
+  }
+
+  Value contributed = getSourceSkipUnary(
+      isa<BlockArgument>(reductionLHS) ? reductionRHS : reductionLHS);
+  Operation *elementwiseOp = contributed.getDefiningOp();
+  if (!elementwiseOp || elementwiseOp->getNumResults() != 1 ||
+      elementwiseOp->getNumOperands() != 2) {
+    llvm::errs() << "expected elementwise op to be binary";
+    return false;
+  }
+  if (!isValidScaledMmaPair(reductionOp, elementwiseOp)) {
+    llvm::errs() << "expected reduction/elementwise op kind not satisfied";
+    return false;
```

Comment:
We should remove those debug prints or make them into proper errors @Muzammiluddin-Syed-ECE